builder/llpcBuilderImplArith.cpp

```diff
@@ -0,0 +1,437 @@
+    // acos coefficient p1 = -0.02363318
+    auto pCoefP1 = GetFPConstant(pInValue->getType(), APFloat(APFloat::IEEEdouble(), APInt(64, 0xBF98334BE0000000)));
+
+    Value* pResult = ASinACosCommon(pInValue, pCoefP0, pCoefP1);
+    return CreateFSub(GetPiByTwo(pResult->getType()), pResult, instName);
+}
+
+// =====================================================================================================================
+// Common code for asin and acos
+Value* BuilderImplArith::ASinACosCommon(
+    Value*        pInValue,   // [in] Input value
+    Constant*     pCoefP0,    // [in] p0 coefficient
+    Constant*     pCoefP1)    // [in] p1 coefficient
+{
+    // asin(x) = sgn(x) * (PI/2 - sqrt(1 - |x|) * (PI/2 + |x| * (PI/4 - 1 + |x| * (p0 + |x| * p1))))
+    // acos(x) = PI/2 - the same, but with slightly different coefficients
+    Value* pAbsInValue = CreateUnaryIntrinsic(Intrinsic::fabs, pInValue);
+    Value* pResult = CreateFMul(pAbsInValue, pCoefP1);
+    pResult = CreateFAdd(pResult, pCoefP0);
+    pResult = CreateFMul(pAbsInValue, pResult);
+    pResult = CreateFAdd(pResult, GetPiByFourMinusOne(pInValue->getType()));
+    pResult = CreateFMul(pAbsInValue, pResult);
+    pResult = CreateFAdd(pResult, GetPiByTwo(pInValue->getType()));
+
+    Value* pSqrtTerm = CreateUnaryIntrinsic(Intrinsic::sqrt,
+                                            CreateFSub(ConstantFP::get(pInValue->getType(), 1.0), pAbsInValue));
+    pResult = CreateFMul(pSqrtTerm, pResult);
+    pResult = CreateFSub(GetPiByTwo(pInValue->getType()), pResult);
+    Value* pSign = CreateFSign(pInValue);
+    return CreateFMul(pSign, pResult);
+}
+
+// =====================================================================================================================
+// Create an "atan" operation for a scalar or vector float or half.
+Value* BuilderImplArith::CreateATan(
+    Value*        pInValue,   // [in] Input value
+    const Twine&  instName)   // [in] Name to give instruction(s)
+{
+    // atan(x) = x - x^3 / 3 + x^5 / 5 - x^7 / 7 + x^9 / 9 - x^11 / 11, |x| <= 1.0
+    // x = min(1.0, x) / max(1.0, x), make |x| <= 1.0
+    Constant* pZero = Constant::getNullValue(pInValue->getType());
+    Constant* pOne = ConstantFP::get(pInValue->getType(), 1.0);
+
+    Value* pAbsX = CreateUnaryIntrinsic(Intrinsic::fabs, pInValue);
+    Value* pMax = CreateBinaryIntrinsic(Intrinsic::maxnum, pAbsX, pOne);
+    Value* pMin = CreateBinaryIntrinsic(Intrinsic::minnum, pAbsX, pOne);
+    Value* pBoundedX = CreateFMul(pMin, CreateFDiv(pOne, pMax));
+    Value* pSquare = CreateFMul(pBoundedX, pBoundedX);
+    Value* pCube = CreateFMul(pSquare, pBoundedX);
+    Value* pPow5 = CreateFMul(pCube, pSquare);
+    Value* pPow7 = CreateFMul(pPow5, pSquare);
+    Value* pPow9 = CreateFMul(pPow7, pSquare);
+    Value* pPow11 = CreateFMul(pPow9, pSquare);
+
+    // coef1 = 0.99997932
```

Comment:
please use native double value, instead of hex code. you can initialize it as following.
union {
    uint64_t u64Val;
    double dVal;
} val;
val.dVal = 0.99997932;

Reply:
1. If we want to use decimal values, it is easier to just use ConstantFP::get(pTy, 0.99997932)
2. But we do not want to use decimal values, as there is no guarantee that different host compilers will give us exactly the same value. That is why I used bit representations in hex.

Reply:
OK, please continue with current implementation.